### DISCUSSION TITLE
Is there any way to set build options for the autograder?

### DISCUSSION THREAD

**Discussion by wes-chen** [gradescope/autograder_samples#70] [Q&A] [answered]
Hi, 

I'm writing an autograder for a computer architecture class using the "Manual Docker Configuration". This involves getting Intel Questa onto a Docker image, and then supplying that image to the autograder. However, there is one caveat. Questa requires a license. For the license that I have generated to work, every image created needs to match a specific mac address that is matched to the license. So the run command would look like:
`sudo docker run -it --mac-address 00:00:00:00:00:00 mydockerautograder:latest bash` for me to get an interactive shell locally.

However, I'm not sure how to translate this `--mac-address` option into Gradescope. I have `ifconfig` included in my run_autograder script, and it seems like the ether address (the mac address) changes for every run. Is there a way for me to include a run option in the image somehow, or tell Gradescope to run my docker image with specific options? My alternative would be to generate licenses for each possible MAC address of the images and then switch the path according to the output of ifconfig, but that feels quite hacky, and I don't know what the possible range of MAC addresses could be.

Please let me know if I'm just missing something!

Thanks.

**Comment by ibrahima** [comment, marked as answer]
Hi @wes-chen! Sorry for the delay in getting back to you! It's not possible to customize the MAC addresses used, however, the alternative approach might work for you.

> My alternative would be to generate licenses for each possible MAC address of the images and then switch the path according to the output of ifconfig, but that feels quite hacky, and I don't know what the possible range of MAC addresses could be.

This is something that other users with similar licensing requirements have done in the past. In our experience the list of MAC addresses used by Docker is pretty fixed. Here's some (third-party) documentation about that: https://macaddress.io/faq/how-to-recognise-a-docker-container-by-its-mac-address

Also, it's helpful to know that IIRC we currently max out at running 16 concurrent containers on a single server instance (depending on specs requested). Based on the above, I believe you should be able to pre-generate licenses for the MAC address range from 02:42:ac:11:00:00 to 02:42:ac:11:00:0F or so (maybe up to :10, I forget if it starts from 0 or 1). You might then need some kind of code within your run_autograder script that choose the appropriate license file to use based on the MAC address at runtime before the licensed software starts.

Hope that helps! Let us know if you have any further questions about this.
